RFC: windowing/gbm: set HDMI and color drm properties

Kwiboo opened the pull request:
## Description
This is a request-for-comment on changes to make use of new drm properties.

This adds work-in-progress code to set the following new properties:
- `content type`: allows userspace to control what content type to use in HDMI infoframes (currently only implemented by Intel driver in v4.19)
- ~~`COLOR_ENCODING` and `COLOR_RANGE`: allows userspace to signal what color encoding is used on a YUV plane (currently only intel and malidp is using this property, v4.18/4.19)~~ #14452

This also adds settings for HDMI output format for use with future drm properties (currently only bsp kernels support YCbCr output signal)

Settings has been duplicated, one for general use and another for when direct-to-plane video is activated.

Please get back with feedback on the code and the added settings.
- Should I use the render flags methods to determin color encoding?
- Should I merge the two different videoscreen and videoplayer settings?

ping @lrusak @FernetMenta @fritsch

## Motivation and Context
Mainline linux is starting to add drm properties that makes it possible to control HDMI signal, add code and settings to control the output signal.

## How Has This Been Tested?
This has not been tested.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

lrusak commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VideoLayerBridgeDRMPRIME.h:
```diff
@@ -39,3 +39,14 @@ class CVideoLayerBridgeDRMPRIME
   CVideoBufferDRMPRIME* m_prev_buffer = nullptr;
 };
+
+enum drm_color_encoding {
```
You might want to make a note about where these come from. Hopefully they are included in the public uapi at some point.

Kwiboo replied:
Sure I will add a comment about origin, they are currently missing in the kernel uapi headers, I expect it will be added in the future.

lrusak commented on xbmc/windowing/gbm/DRMAtomic.cpp:
```diff
@@ -119,6 +111,14 @@ void CDRMAtomic::FlipPage(struct gbm_bo *bo, bool rendered, bool videoLayer)
     }
   }
 
+  uint32_t flags = 0;
```
why is this commit needed?

Kwiboo replied:
I should probably add a better commit comment :)

I noticed a one time error with the `CDRMUtils::DrmFbGetFromBo(bo)` call when I was testing rockchip HDR stuff some weeks ago.
Looking at the code the `m_need_modeset` is reset on such error and next flip there is no modeset allowed.
With this change the modeset is just delayed until `DrmFbGetFromBo` is successful.

lrusak replied:
makes sense. I just wanted to know the reason behind the change

lrusak reviewed (commented):
Looks good so far! I will hopefully give it a test after work today!

I am ok with having two separate settings for the GUI and Videoplayer for now. If it feels clunky later then we can change it.

I would just add the settings visibility based on if the DRM property exists then it should be good to go 👍

a1rwulf commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VideoLayerBridgeDRMPRIME.cpp:
```diff
@@ -140,6 +166,14 @@ void CVideoLayerBridgeDRMPRIME::Configure(CVideoBufferDRMPRIME* buffer)
+  {
+    m_DRM->AddProperty(plane, "COLOR_ENCODING", GetColorEncoding(buffer));
+    m_DRM->AddProperty(plane, "COLOR_RANGE", GetColorRange(buffer));
+  }
```
Not sure if I understand this code?
You check if the property is there and if it is you add another one?

lrusak replied:
@a1rwulf we need to check if the drm driver has the property available. If it is available then we can set the property. Maybe the wording is just odd.

a1rwulf replied:
yeah maybe I'm confused because we have AddProperty and SetProperty in our public API in DRMUtils.

a1rwulf replied:
Maybe we can make DRMLegacy override AddProperty instead of SetProperty and just remove the SetProperty method?

lrusak replied:
Ah, it looks like it's just wording between legacy (`drmModeObjectSetProperty`) and atomic (`drmModeAtomicAddProperty`) operations. Maybe we can clean that up at some point.

Kwiboo replied:
`HasProperty` checks if the plane supports the property or not and the `AddProperty` adds a new property value to the next commit.
Should I rename `HasProperty` to `SupportsProperty` to make it less confusing?

The reason for the `HasProperty` check is to not trigger warnings when the plane don't support the property.

a1rwulf replied:
That would be nice.
Still we could also cleanup SetProperty vs. AddProperty as well.

a1rwulf commented:
> Should I merge the two different videoscreen and videoplayer settings?

If technically possible I'd vote for merge.
But I guess there's a reason why you duplicated.

Kwiboo commented:
> If technically possible I'd vote for merge.
> But I guess there's a reason why you duplicated.

The reason I went with duplicate settings is to be able to use RGB signal and Game content type when navigating or gaming, and then switch to YCbCr and Cinema/Graphics when a video is started.
That should theoretically allow for sending untouched YUV video to TV, and still have best possible image/colors when gaming or watching photos.

lrusak commented:
@Kwiboo after closer inspection I cannot see how this sets the color range for the gui. It looks like you've only implemented it for the DRM prime renderer?

Also I think we need to implement some settings callbacks so we can toggle the setting without restarting

Kwiboo commented:
@lrusak @a1rwulf I have updated this with `ISettingCallback` support and added a `HDMI Quantization Range` setting for use with Intel Broadcast RGB property, still unsure if this should be added as it is an Intel specifc property.

`CDRMAtomic::OnSettingChanged` can probably be simplified to a oneliner `m_need_modeset = true;`

lrusak commented:
@Kwiboo looks good! my only quarrel is that the limited range function should use the existing setting not add a new one IMO.

a1rwulf commented:
> @Kwiboo looks good! my only quarrel is that the limited range function should use the existing setting not add a new one IMO.

Given the summary of our wiki article about video levels there are 3 options that give correct results:

> Full, Full, Full -- Video Level CORRECT : WtW & BtW Clipped, Single scaling
> Full, Full, Limited -- Video Level WRONG : Crushed Blacks
> Full, Limited, Full -- Video Level WRONG : Washed Out Colors
> Full, Limited, Limited -- Video Level CORRECT : WtW & BtW Clipped, Double scaling
> Limited, Full, Full -- Video Level WRONG : Washed Out Colors
> Limited, Full, Limited -- Video Level CORRECT : WtW & BtB preserved, Zero scaling
> Limited, Limited, Full -- Video Level WRONG : Extreme Washed Out Colors
> Limited, Limited, Limited -- Video Level WRONG : Washed Out Colors

So "Full, Full, Full" and "Full, Limited, Limited" could be used when kodi is put to full.
Imho we could scrap the Full, Limited, Limited option and go for wether:
Full, Full, Full or Limited, Full, Limited.

That would mean we could use the existing setting and put the driver to Full always.
What do you think about this idea?

Rechi commented:
jenkins build this please

Kwiboo committed:
windowing/gbm: set Intel Broadcast RGB connector property

Kwiboo committed:
windowing/gbm: add setting for HDMI Output Format

Kwiboo committed:
windowing/gbm: add setting for HDMI Quantization Range

Kwiboo committed:
windowing/gbm: set HDMI connector content type property

Kwiboo committed:
windowing/gbm: add GetConnector method

phunkyfish commented:
@Kwiboo I assume you are still interested in merging this? What needs to be done to progress it?

phunkyfish commented:
As there has been no comment on this PR it will be closed in 2 weeks time if no comment stating to do otherwise is added.

Kwiboo commented:
There is still no common linux drm connetor property for configuring HDMI Output Format or HDMI Quantization Range. HDMI Content Type property is part of mainline linux at the moment.

I will close this for now, it can be re-opened / re-submitted when drm connector properties are closer to become part of mainline linux.
